src/aleph/storage.py

```diff
@@ -22,113 +24,207 @@
 LOGGER = logging.getLogger("STORAGE")
 
 
-async def json_async_loads(s: str):
+class ContentSource(str, Enum):
+    DB = "DB"
+    P2P = "P2P"
+    IPFS = "IPFS"
+    INLINE = "inline"
+
+
+@dataclass
+class StoredContent:
+    hash: str
+    source: Optional[ContentSource]
+
+
+@dataclass
+class RawContent(StoredContent):
+    value: bytes
+
+    def __len__(self):
+        return len(self.value)
+
+
+@dataclass
+class JsonContent(StoredContent):
+    value: Any
+    raw_content: bytes
+
+
+async def json_async_loads(s: AnyStr):
     """Deserialize ``s`` (a ``str``, ``bytes`` or ``bytearray`` instance
     containing a JSON document) to a Python object in an asynchronous executor."""
     return await run_in_executor(None, json.loads, s)
 
 
-async def get_message_content(message: Dict):
+async def get_message_content(message: Dict) -> JsonContent:
     item_type: str = message.get("item_type", ItemType.IPFS)
+    item_hash = message["item_hash"]
 
     if item_type in (ItemType.IPFS, ItemType.Storage):
-        return await get_json(message["item_hash"], engine=ItemType(item_type))
+        return await get_json(item_hash, engine=ItemType(item_type))
     elif item_type == ItemType.Inline:
         if "item_content" not in message:
-            LOGGER.warning(f"No item_content in message {message.get('item_hash')}")
-            return -1, 0  # never retry, bogus data
+            error_msg = f"No item_content in message {message.get('item_hash')}"
+            LOGGER.warning(error_msg)
+            raise InvalidContent(error_msg)  # never retry, bogus data
         try:
             item_content = await json_async_loads(message["item_content"])
-        except (json.JSONDecodeError, KeyError):
-            LOGGER.exception("Can't decode JSON")
-            return -1, 0  # never retry, bogus data
-        return item_content, len(message["item_content"])
+        except (json.JSONDecodeError, KeyError) as e:
+            error_msg = f"Can't decode JSON: {e}"
+            LOGGER.exception(error_msg)
```

Comment:
Isn't this already logged by the caller ?

Reply:
No, not necessarily and not with that level of detail. I'm not sure it should be logged at exception level though, it's not an issue with our code but with the content of the message.